### ISSUE TITLE
Support exhausted `match` statement with several enums

### ISSUE THREAD

**Issue by simensol** [microsoft/pyright#3073]
**Describe the bug**
Using a `match` statement with two or more enums raises `reportMatchNotExhaustive` although all values are exhaustively handled.

**To Reproduce**

```python
from enum import Enum


class First(Enum):
    A = 1
    B = 2


class Seccond(Enum):
    C = 1
    D = 2


def test1(first: First, seccond: Seccond):
    match (first, seccond):
        case (First.A | First.B, Seccond.C):
            pass
        case (First.A | First.B, Seccond.D):
            pass


def test2(first: First, seccond: Seccond):
    match (first, seccond):
        case (First.A, Seccond.C):
            pass
        case (First.A, Seccond.D):
            pass
        case (First.B, Seccond.C):
            pass
        case (First.B, Seccond.D):
            pass
```

Pyright raises `reportMatchNotExhaustive` for both `test1()` and `test2()`:

```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "tuple[First, Seccond]"
  If exhaustive handling is not intended, add "case _: pass" Pylance(reportMatchNotExhaustive)
```

Of course, if I add `case _: pass` the error disappears. However, if I later add a new value to either the `First` or `Second` enums, Pyright will not warn me that the new value isn't handled – it is swallowed by `case _: pass`.

**Expected behavior**
I expect Pyright not to raise `reportMatchNotExhaustive` when all possible combinations of enum values are handled by the `match` statement.

**Screenshots**
<img width="363" alt="tests_py_—_backend__Dev_Container__backend_" src="https://user-images.githubusercontent.com/4491653/154672399-e516f865-6034-4bf8-9648-dd012061153e.png">

**VS Code extension or command-line**
Pylance v2022.2.3.

**Comment by erictraut** [comment]
Sorry, but this is unlikely to happen. This would involve narrowing combinations of types. The combinatorics would be be very expensive, and it would require a very different type narrowing architecture than the one used in pyright.

I recommend using a nested `match` statement instead.

**Comment by anderslundstedt** [comment]
What about supporting the special case of handling tuples of the builtin enum `bool`? That would be quite helpful—comparing test1.py and test2.py below, test1.py is a lot easier to parse (for me at least).

```
% pyright --version
pyright 1.1.275
```

`test1.py` in an otherwise empty temporary directory:
```
# pyright: strict

def bb() -> tuple[bool,bool]:
    return (False, True)

match bb():
    case (True,   True): print('tt')
    case (True,  False): print('tf')
    case (False,  True): print('ft')
    case (False, False): print('ff')
```

```
% python3.10 test1.py
ft
```

```
% pyright test1.py
No configuration file found.
No pyproject.toml file found.
stubPath /private/var/folders/0s/vqqb3xgn0kl45344k6v85d180000gn/T/tmp.0JBo3oncJb/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.275
/private/var/folders/0s/vqqb3xgn0kl45344k6v85d180000gn/T/tmp.0JBo3oncJb/test1.py
  /private/var/folders/0s/vqqb3xgn0kl45344k6v85d180000gn/T/tmp.0JBo3oncJb/test1.py:6:7 - error: Cases within match statement do not exhaustively handle all values
    Unhandled type: "tuple[bool, bool]"
    If exhaustive handling is not intended, add "case _: pass" (reportMatchNotExhaustive)
1 error, 0 warnings, 0 informations 
Completed in 0.834sec
```

test2.py in an otherwise empty temporary directory:
```
# pyright: strict

def bb() -> tuple[bool,bool]:
    return (True, False)

b1, b2 = bb()
match b1:
    case True:
        match b2:
            case True:  print('tt')
            case False: print('tf')
    case False:
        match b2:
            case True:  print('ft')
            case False: print('ff')
```

```
% python3.10 test2.py
tf
```

```
% pyright test2.py
No configuration file found.
No pyproject.toml file found.
stubPath /private/var/folders/0s/vqqb3xgn0kl45344k6v85d180000gn/T/tmp.FyDCQdV8QO/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.275
0 errors, 0 warnings, 0 informations 
Completed in 0.832sec
```

**Comment by erictraut** [comment]
@anderslundstedt, this issue has been closed as completed for a long time. If you want to propose new features, please open a new issue. What you're suggesting has already been suggested, but it's unlikely to be implemented. It would require a form of type narrowing that is not used anywhere else in any type checker that I'm aware of. It would require that individual typed elements within a tuple be narrowed. That's theoretically possible, but quite complex and probably not worth it for this particular situation. If you want to take advantage of exhausted match checking, you can change your code slightly to work with the current type narrowing capabilities.
